### PR TITLE
Allow dashes according to RFC in all domain parts except TLD

### PR DESCRIPTION
**Pull request by Jikstra**
Fixes https://github.com/markdown-it/linkify-it/issues/63

### CODE REVIEW

**Comment by puzrin** [comment]
- https://github.com/markdown-it/linkify-it/issues/63#issuecomment-393289287 according to discussion, no options needed for correct implementation.
- Test patterns should be placed in https://github.com/markdown-it/linkify-it/masterf/test/fixtures/links.txt when possible. Also, the real domain names required, not artificial ones.

**Comment by Jikstra** [comment]
> #63 (comment) according to discussion, no options needed for correct implementation.

So you want this to be the default behaviour? I'm not too familiar with markdown and it's limitiations, so i don't know when a url with multiple dashes would cause trouble.

> Test patterns should be placed in https://github.com/markdown-it/linkify-it/masterf/test/fixtures/links.txt when possible. Also, the real domain names required, not artificial ones.

Thanks, will try to update the code accordingly.

**Comment by Jikstra** [comment]
@puzrin on more question, is `http://a.b--c.de/   % `--` disabled, because collision possible` still a url which shouldn't be allowed? Because writing a regex which checks that the dashes are only in the subdomain part is maybe a lot harder than just removing the small regex part which i made optional with the opts flag.

**Comment by puzrin** [comment]
> So you want this to be the default behaviour?

Yes.

> I'm not too familiar with markdown and it's limitiations, so i don't know when a url with multiple dashes would cause trouble.

If you do exactly as discussed in #63, that will be ok. We already discussed the best possible behaviour. Any deviation will require to do all checks again.

**Comment by puzrin** [comment]
New relaxed rule should touch only domain names (except TLD), and should not touch path.

**Comment by Jikstra** [comment]
My fix also allows urls like `https://net--lify.com` which you can't register (so i guess they are invalid). There was the one not-link test which i removed, i don't think this is a problem (and if so only a collission with markdown). If this is really a problem, i'm going to try to implement a regex which only allows multiple dashes in the second or lower levels.

File: lib/re.js [hunk 1]
```diff
@@ -101,10 +101,8 @@ module.exports = function (opts) {
       '|' +
       '(?:' + re.src_pseudo_letter + ')' +
       '|' +
-      // don't allow `--` in domain names, because:
-      // - that can conflict with markdown &mdash; / &ndash;
-      // - nobody use those anyway
-      '(?:' + re.src_pseudo_letter + '(?:-(?!-)|' + re.src_pseudo_letter + '){0,61}' + re.src_pseudo_letter + ')' +
+       '(?:' + re.src_pseudo_letter + '(?:-' +
+        '|' + re.src_pseudo_letter + '){0,61}' + re.src_pseudo_letter + ')' +
```
Comment (puzrin) [inline lib/re.js]: This 2 lines have broken formatting/intent - not like other code. Please, follow conding style for other parts of file. All you had to do - just remove `(?!-)` and don't touch anything else.

File: test/fixtures/links.txt [hunk 2]
```diff
@@ -302,3 +302,9 @@ test@google.com
 
 ｜http://google.com｜bar
 http://google.com
+
+%
+% Domains with multiple dashes
+%
+
+https://5b0ee223b312746c1659db3f--thelounge-chat.netlify.com/docs/
```
Comment (puzrin) [inline test/fixtures/links.txt]: Missed other samples, discussed in #63.

**Comment by Jikstra** [comment]
hey @puzrin is anything still blocking you from merging this?

**Comment by xPaw** [comment]
Thanks @Jikstra!

Hopefully a package update can be pushed to npm too.

**Comment by Jikstra** [comment]
@xPaw no problem, the-lounge seems nice, know some people who are using it. I'm glad i could helped you :)

**Pull request merged**